modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/ExecutionTest.java

```diff
@@ -1046,6 +1047,36 @@ else if (o2 != null)
         }
     }
 
+    /**
+     * Test verifies that an AssertionError thrown from an execution node
+     * proprely handled by a task executor.
+     */
+    @Test
+    @SuppressWarnings("ThrowableNotThrown")
+    public void assertionHandlingTest() {
+        ExecutionContext<Object[]> ctx = executionContext(F.first(nodes()), UUID.randomUUID(), 0);
+        IgniteTypeFactory tf = ctx.getTypeFactory();
+        RelDataType rowType = TypeUtils.createRowType(tf, int.class, String.class);
+
+        CorruptedNode<Object[]> node = new CorruptedNode<>();
+
+        RootNode<Object[]> root = new RootNode<>(ctx, rowType);
+        root.register(node);
+
+        new Thread(() -> {
```

Comment:
don`t understand the purpose of this nonjoining thread. If u want to close the node - close it in finally

Reply:
before this patch caller thread got staled on `root.hasNext()` invocation, thus we need someone who explicitly cancels an execution, because in our case the caller thread is a thread that runs the test